runtime/gc_ctrl.c

```diff
@@ -293,6 +293,23 @@ CAMLprim value caml_gc_major_slice (value v)
 }
 
 CAMLprim value caml_gc_compaction(value v)
+{
+  Caml_check_caml_state();
+  CAML_EV_BEGIN(EV_EXPLICIT_GC_COMPACT);
+  CAMLassert (v == Val_unit);
+  value exn = gc_major_exn(0);
+  if (Is_exception_result(exn))
+    return caml_raise_if_exception(exn);
+  exn = gc_major_exn(0);
+  if (Is_exception_result(exn))
+    return caml_raise_if_exception(exn);
+  exn = gc_major_exn(1);
```

Comment:
Two questions:
- why is the code so different from `gc_full_major_exn` above? could one share code between the two?
- on the caller side, it is very unclear that `gc_major_exn(0)` means "do a major GC cycle" while `gc_major_exn(1)` means "do a compaction". Could there be a clearer API for non-experts reading the code? You could have an enumeration with proper names, or even two different entry points to use whenever the flag value is a constant / statically known.

Reply:
I've reworked the code so it matches `gc_full_major_exn`.

Agreed on changes to the API to make it more obvious what's happening. To be clear, `gc_major_exn(0)` means "do a major GC cycle", `gc_major_exn(1)` means "do a major GC cycle that includes a compaction as well".

We should create an issue and tackle in another PR.